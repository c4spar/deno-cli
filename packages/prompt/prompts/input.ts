import { blue } from 'https://deno.land/std@0.63.0/fmt/colors.ts';
import { Figures } from '../lib/figures.ts';
import { GenericInput, GenericInputKeys, GenericInputPromptOptions, GenericInputPromptSettings } from '../lib/generic-input.ts';

export interface InputKeys extends GenericInputKeys {}

export interface InputOptions extends GenericInputPromptOptions<string> {
    minLength?: number;
    maxLength?: number;
    keys?: InputKeys;
}

interface InputSettings extends GenericInputPromptSettings<string> {
    minLength: number;
    maxLength: number;
    keys?: InputKeys;
}

export class Input extends GenericInput<string, InputSettings> {

    public static async prompt( options: string | InputOptions ): Promise<string> {

        if ( typeof options === 'string' ) {
            options = { message: options };
        }

        return new this( {
            pointer: blue( Figures.POINTER_SMALL ),
            minLength: 0,
            maxLength: Infinity,
            ...options
        } ).prompt();
    }

    protected validate( value: string ): boolean | string {
        if ( typeof value !== 'string' ) {
            return false;
        }
        if ( value.length < this.settings.minLength ) {
            return `Value must be longer then ${ this.settings.minLength } but has a length of ${ value.length }.`;
        }
        if ( value.length > this.settings.maxLength ) {
            return `Value can't be longer then ${ this.settings.maxLength } but has a length of ${ value.length }.`;
        }
        return true;
    }

    protected transform( value: string ): string | undefined {
        return value.trim();
    }

    protected format( value: string ): string {
        return value;
    }
}
