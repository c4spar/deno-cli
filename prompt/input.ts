import { blue } from "./deps.ts";
import { Figures } from "./figures.ts";
import {
  GenericInput,
  GenericInputKeys,
  GenericInputPromptOptions,
  GenericInputPromptSettings,
} from "./_generic_input.ts";

export type InputKeys = GenericInputKeys;

/** Input prompt options. */
export interface InputOptions extends GenericInputPromptOptions<string> {
  minLength?: number;
  maxLength?: number;
  keys?: InputKeys;
}

/** Input prompt settings. */
interface InputSettings extends GenericInputPromptSettings<string> {
  minLength: number;
  maxLength: number;
  keys?: InputKeys;
}

/** Input prompt representation. */
export class Input extends GenericInput<string, InputSettings> {
  /** Execute the prompt and show cursor on end. */
  public static async prompt(options: string | InputOptions): Promise<string> {
    if (typeof options === "string") {
      options = { message: options };
    }

    return new this({
      pointer: blue(Figures.POINTER_SMALL),
      minLength: 0,
      maxLength: Infinity,
      ...options,
    }).prompt();
  }

  /**
   * Validate input value.
   * @param value User input value.
   * @return True on success, false or error message on error.
   */
  protected validate(value: string): boolean | string {
    if (typeof value !== "string") {
      return false;
    }
    if (value.length < this.settings.minLength) {
      return `Value must be longer then ${this.settings.minLength} but has a length of ${value.length}.`;
    }
    if (value.length > this.settings.maxLength) {
      return `Value can't be longer then ${this.settings.maxLength} but has a length of ${value.length}.`;
    }
    return true;
  }

  /**
   * Map input value to output value.
   * @param value Input value.
   * @return Output value.
   */
  protected transform(value: string): string | undefined {
    return value.trim();
  }

  /**
   * Format output value.
   * @param value Output value.
   */
  protected format(value: string): string {
    return value;
  }
}
