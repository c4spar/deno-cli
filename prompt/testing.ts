import { eraseDown } from "../ansi/ansi_escapes.ts";
import { basename } from "./deps.ts";
import { assertSnapshot } from "https://deno.land/std@0.170.0/testing/snapshot.ts";

export interface AssertPromptSnapshotOptions {
  meta: ImportMeta;
  tests: Record<string, Array<string>>;
  osSuffix?: Array<typeof Deno.build.os>;
  fn(): Promise<unknown>;
}

const encoder = new TextEncoder();

/** Run prompt snapshot tests. */
export function assertPromptSnapshot(
  options: AssertPromptSnapshotOptions,
): Promise<void>;

/** Run prompt snapshot tests. */
export function assertPromptSnapshot(
  meta: ImportMeta,
  tests: Record<string, Array<string>>,
  fn: () => Promise<unknown>,
): Promise<void>;

/**
 * Run prompt snapshot tests.
 *
 * ```ts
 * import { assertPromptSnapshot } from "./testing.ts";
 * import { Input } from "./input.ts";
 *
 * await assertPromptSnapshot({
 *   meta: import.meta,
 *   tests: {
 *     "should enter som text": ["foo bar", "\n"],
 *   },
 *   async fn() {
 *     await Input.prompt({
 *       message: "Whats your name?",
 *       default: "foo",
 *     });
 *   },
 * });
 * ```
 *
 * @param optionsOrMeta import.meta
 * @param tests         Object of tests. Key is the test name and the value is
 *                      an array of input sequences/characters.
 * @param fn            Test function.
 */
export async function assertPromptSnapshot(
  optionsOrMeta: AssertPromptSnapshotOptions | ImportMeta,
  tests?: Record<string, Array<string>>,
  fn?: () => Promise<unknown>,
): Promise<void> {
  const options: AssertPromptSnapshotOptions = "meta" in optionsOrMeta
    ? optionsOrMeta
    : {
      meta: optionsOrMeta,
      tests: tests!,
      fn: fn!,
    };

  if (options.meta.main) {
    await options.fn();
  } else {
    registerTest(options);
  }
}

function registerTest(options: AssertPromptSnapshotOptions) {
  const fileName = basename(options.meta.url);

  Deno.test({
    name: `assert prompt snapshot - ${fileName}`,
    async fn(ctx) {
      const tests = Object.entries(options.tests ?? {});
      if (!tests.length) {
        throw new Error(`No tests defined for: ${options.meta.url}`);
      }

      for (const [name, inputs] of tests) {
        await ctx.step({
          name,
          async fn() {
            const output: string = await runPrompt(options.meta.url, inputs);
            const suffix = options.osSuffix?.includes(Deno.build.os)
              ? `.${Deno.build.os}`
              : "";

            await assertSnapshot(ctx, output, {
              path: `__snapshots__/${fileName}${suffix}.snap`,
            });
          },
        });
      }
    },
  });
}

async function runPrompt(
  url: string,
  inputs: Array<string>,
): Promise<string> {
  const cmd = new Deno.Command("deno", {
    stdin: "piped",
    stdout: "piped",
    stderr: "piped",
    args: [
      "run",
      "--allow-all",
      url,
    ],
    env: {
      NO_COLOR: "true",
    },
  });
  const child: Deno.ChildProcess = cmd.spawn();
  const writer = child.stdin.getWriter();

  for (const input of inputs) {
    await writer.write(encoder.encode(input));
    // Ensure all inputs are processed and rendered separately.
    await new Promise((resolve) => setTimeout(resolve, 600));
  }

  const { success, stdout, stderr } = await child.output();
  writer.releaseLock();
  await child.stdin.close();

  if (!success) {
    throw new Error(`test failed: ${url}`);
  }

  // Add a line break after each test input.
  return "stdout:\n" + new TextDecoder().decode(stdout).replaceAll(
    eraseDown(),
    eraseDown() + "\n",
  ) + "\nstderr:\n" + new TextDecoder().decode(stderr).replaceAll(
    eraseDown(),
    eraseDown() + "\n",
  );
}
