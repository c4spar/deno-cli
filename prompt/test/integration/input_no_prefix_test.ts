import { Input } from "../../input.ts";
import { assertPromptSnapshot } from "../../testing.ts";

await assertPromptSnapshot({
  name: "input prompt without prefix",
  meta: import.meta,
  osSuffix: ["windows"],
  steps: {
    "should disable prefix": ["bar", "\n"],
  },
  async fn() {
    await Input.prompt({
      message: "Whats your name?",
      default: "foo",
      prefix: "",
    });
  },
});
