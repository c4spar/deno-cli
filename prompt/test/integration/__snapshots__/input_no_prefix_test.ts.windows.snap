export const snapshot = {};

snapshot[`prompt snapshot tests file:///Users/c4spar/workspace/deno/deno-cliffy/prompt/test/integration/input_no_prefix_test.ts 1`] = `
"stdout:
Whats your name? (foo) » \\x1b[26G\\x1b[G\\x1b[0J
Whats your name? (foo) » bar\\x1b[29G\\x1b[G\\x1b[0J
Whats your name? (foo) » bar
\\x1b[?25h\\x1b[?25h
stderr:
"
`;
